working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,612 @@
+
+<br />
+
+<details>
+  <summary>
+    Try out adding error handling and updating the test assertions, then expand this section to see how we changed <code>update_book</code>, <code>test_update_book_without_title</code>, and <code>test_update_book_without_description</code>.
+  </summary>
+
+  **Updated `update_book` function:**
+  ```py
+  @books_bp.route("/<book_id>", methods=["PUT"])
+  def update_book(book_id):
+      book = validate_book(book_id)
+
+      request_body = request.get_json()
+
+      try:
+          book.title = request_body["title"]
+          book.description = request_body["description"]
+
+      except KeyError as e:
+          abort(make_response({"message": f"{e.args[0]} is required"}, 400))
+
+      db.session.commit()
+
+      return make_response(jsonify(f"Book #{book.id} successfully updated"))
+  ```
+
+  **Updated tests:**
+  ```py
+  def test_update_book_without_title(client, two_saved_books):
+      # Arrange
+      book_id = 1
+      data = {
+          "description": "New description"
+      }
+
+      # Act
+      response = client.put(f"/books/{book_id}", json=data)
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 400
+      assert response_body == {"message": "title is required"}
+
+  def test_update_book_without_description(client, two_saved_books):
+      # Arrange
+      book_id = 1
+      data = {
+          "title": "Updated Book"
+      }
+
+      # Act
+      response = client.put(f"/books/{book_id}", json=data)
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 400
+      assert response_body == {"message": "description is required"}
+  ```
+</details>
+
+Now that we've seen several facets of working with Copilot to write tests, we will leave it as a personal exercise to complete the test suite for our last untested dependency `delete_book`.
+
+#### Executing the Refactor
+
+Now that we have a solid test suite, we can get some help from Copilot to make this refactor go a little more quickly. Let's review what needs to change in `validate_book` to make the function more flexible so we can use it with other models:
+- take in a second parameter that represents a model class reference
+- change any uses of the `Book` class to use our new parameter
+- update the function name, variable names, and return messages so they reflect that the function is not specific to the `Book` class
+
+The changes we need are pretty minimal and could be quick to perform manually, but for the practice, let's see how Copilot handles the change. Let's highlight the `validate_book` function, open up the inline Copilot chat, and provide the following prompt:
```

Comment:
- fine as is, but on first read, "the" tripped me up. I think I interpret leaving off "the" as more of an introduction to what we'll do, while "for the practice" more answers the question "why are we doing something?"
```suggestion
The changes we need are pretty minimal and could be quick to perform manually, but for practice, let's see how Copilot handles the change. Let's highlight the `validate_book` function, open up the inline Copilot chat, and provide the following prompt:
```